### ISSUE TITLE
 alpaka math functions does not work in lambdas with alpaka 0.9 and CUDA backend

### ISSUE THREAD

**Issue by SimeonEhrig** [alpaka-group/alpaka#1688]
# Problem

I want to pass a lambda to a kernel, which is called and use a `alpaka::math` function. Therefore the lambda needs an `acc` object as argument. The CUDA backend does not support generic lambdas, therefore it needs to be an explicit type. This causes a compiler error. 

```bash
FAILED: CMakeFiles/alpaka_lambda.dir/main.cpp.o 
/opt/spack-modules/linux-ubuntu20.04-skylake_avx512/gcc-10.2.0/cuda-11.0.2-mtxiuzc35ykbmjk4fnpmhwnjbgquvcv5/bin/nvcc -forward-unknown-to-host-compiler -ccbin=/usr/bin/g++ -DALPAKA_ACC_GPU_CUDA_ENABLED -DALPAKA_BLOCK_SHARED_DYN_MEMBER_ALLOC_KIB=47 -DALPAKA_DEBUG=0 -DALPAKA_DEBUG_OFFLOAD_ASSUME_HOST -DALPAKA_OFFLOAD_MAX_BLOCK_SIZE=256 -DBOOST_ALL_NO_LIB -isystem=/opt/spack-modules/linux-ubuntu20.04-skylake_avx512/gcc-10.2.0/cuda-11.0.2-mtxiuzc35ykbmjk4fnpmhwnjbgquvcv5/targets/x86_64-linux/include -isystem=/home/sehrig/application/miniconda3/envs/vikunja-dev/include -isystem=/opt/spack-modules/linux-ubuntu20.04-skylake_avx512/gcc-10.2.0/boost-1.74.0-lqyr7ou7iafiwvq5fb6ipp62o6l2gqlf/include -isystem=/opt/spack-modules/linux-ubuntu20.04-skylake_avx512/gcc-10.2.0/cuda-11.0.2-mtxiuzc35ykbmjk4fnpmhwnjbgquvcv5/include --generate-code=arch=compute_52,code=[compute_52,sm_52] -Xcompiler=-fPIE --extended-lambda --expt-relaxed-constexpr -Xcudafe=--display_error_number -Xcudafe=--diag_suppress=esa_on_defaulted_function_ignored -Xcompiler -pthread -std=c++17 -MD -MT CMakeFiles/alpaka_lambda.dir/main.cpp.o -MF CMakeFiles/alpaka_lambda.dir/main.cpp.o.d -x cu -c ../main.cpp -o CMakeFiles/alpaka_lambda.dir/main.cpp.o
/home/sehrig/application/miniconda3/envs/vikunja-dev/include/alpaka/math/Traits.hpp(850): error #3050: calling a __device__ function("operator()") from a __host__ __device__ function("max") is not allowed
          detected during instantiation of "auto alpaka::math::max(const T &, const Tx &, const Ty &) [with T=alpaka::AccGpuCudaRt<alpaka::DimInt<1UL>, std::size_t>, Tx=float, Ty=float]" 
../main.cpp(40): here

/home/sehrig/application/miniconda3/envs/vikunja-dev/include/alpaka/math/Traits.hpp(850): error #3050: calling a __device__ function("operator()") from a __host__ __device__ function("max") is not allowed
          detected during instantiation of "auto alpaka::math::max(const T &, const Tx &, const Ty &) [with T=alpaka::AccGpuCudaRt<alpaka::DimInt<1UL>, std::size_t>, Tx=float, Ty=float]" 
../main.cpp(40): here

2 errors detected in the compilation of "../main.cpp".
ninja: build stopped: subcommand failed.
```

The same code is working with the serial backend, OpenMP backend and also the Hip backend. Maybe it also happen with specific CUDA versions. CUDA 11.0 failed and CUDA 11.2 was fine.

# Code to reproduce

Use the alpaka version of commit [ebc1171](https://github.com/alpaka-group/alpaka/commit/ebc1171feac21f1e21c49bcd9f053e7b01b584d0) (branch `0.9.0-rc1`).

CMakeLists.txt
```cmake
cmake_minimum_required(VERSION 3.18)

project(alpaka_lambda)
set(CMAKE_POSITION_INDEPENDENT_CODE ON)
set(alpaka_CUDA_EXPT_EXTENDED_LAMBDA ON)

option(alpaka_ACC_GPU_CUDA_ENABLE "enable alpaka CUDA accelerator" ON)

find_package(alpaka)

alpaka_add_executable(${PROJECT_NAME} main.cpp)
target_link_libraries(${PROJECT_NAME} PRIVATE alpaka::alpaka)
```

main.cpp
```c++
#include <alpaka/alpaka.hpp>


struct GlobalKernel
{
    template<typename TAcc, typename TFunctor>
    ALPAKA_FN_ACC auto operator()(TAcc const& acc, TFunctor const& functor) const -> void
    {
        float f = functor(acc, 42.f);
        printf("%f\n", f);
    }
};

int main()
{
    using Dim = alpaka::DimInt<1>;
    using Idx = std::size_t;

    using Acc = alpaka::AccGpuCudaRt<Dim, Idx>;
    using Queue = alpaka::Queue<Acc, alpaka::Blocking>;

    auto const devAcc = alpaka::getDevByIdx<Acc>(0u);
    Queue queue(devAcc);

    using Vec = alpaka::Vec<Dim, Idx>;
    Vec const elementsPerThread(Vec::all(static_cast<Idx>(1)));
    Vec const threadsPerGrid(Vec::all(static_cast<Idx>(1)));
    using WorkDiv = alpaka::WorkDivMembers<Dim, Idx>;
    WorkDiv const workDiv = alpaka::getValidWorkDiv<Acc>(
        devAcc,
        threadsPerGrid,
        elementsPerThread,
        false,
        alpaka::GridBlockExtentSubDivRestrictions::Unrestricted);

    // does not work
    auto functor = [] ALPAKA_FN_HOST_ACC(alpaka::AccGpuCudaRt<Dim, Idx> const& acc, float const i) -> float
    {
        // return i + 1.f;
        return alpaka::math::max(acc, 41.f, i) + 0.5f;
    };
    GlobalKernel kernel;

    alpaka::exec<Acc>(queue, workDiv, kernel, functor);
    alpaka::wait(queue);

    return 0;
}
```

# Possible solution

@psychocoderHPC points me, that is required that the CUDA implementation of `max` needs an explicit return type.
I fixed the type of this function to `float` and looks like, that it works. 
https://github.com/alpaka-group/alpaka/blob/b074b0df68a96321dc73261ab2b9d3d41180f18c/include/alpaka/math/MathUniformCudaHipBuiltIn.hpp#L642

**Comment by SimeonEhrig** [comment]
Using a wrapper function for the type deduction also does not work.

```c++
//! The CUDA max trait specialization.
        template<typename Tx, typename Ty>
        struct Max<
            MaxUniformCudaHipBuiltIn,
            Tx,
            Ty,
            std::enable_if_t<std::is_arithmetic_v<Tx> && std::is_arithmetic_v<Ty>>>
        {
    private:
            __device__ auto foo(Tx const& x, Ty const& y)
            {
                if constexpr(std::is_integral_v<Tx> && std::is_integral_v<Ty>)
                    return ::max(x, y);
                else if constexpr(is_decayed_v<Tx, float> && is_decayed_v<Ty, float>)
                    return ::fmaxf(x, y);
                else if constexpr(
                    is_decayed_v<
                        Tx,
                        double> || is_decayed_v<Ty, double> || (is_decayed_v<Tx, float> && std::is_integral_v<Ty>)
                    || (std::is_integral_v<Tx> && is_decayed_v<Ty, float>) )
                    return ::fmax(x, y);
                else
                    static_assert(!sizeof(Tx), "Unsupported data type");

                using Ret [[maybe_unused]] = std::conditional_t<
                    std::is_integral_v<Tx> && std::is_integral_v<Ty>,
                    decltype(::max(x, y)),
                    std::conditional_t<is_decayed_v<Tx, float> && is_decayed_v<Ty, float>, float, double>>;
                ALPAKA_UNREACHABLE(Ret{});
            }
    public:
      __device__ auto operator()(MaxUniformCudaHipBuiltIn const& /* max_ctx */, Tx const& x, Ty const& y) -> decltype(foo(x,y))
      {
        return foo(x, y);
      }
};
```

**Comment by psychocoderHPC** [comment]
In the past we have seen nearly an equal issue and solved it via explicit definition of the return type: see https://github.com/alpaka-group/alpaka/pull/1085

**Comment by SimeonEhrig** [comment]
I test the example with CUDA 11.0, 11.1, 11.2 and 11.6 and the same error appears in all versions.

**Comment by j-stephan** [comment]
If this happens in CUDA 11.6 we should open a NVIDIA bug report. Can you create a reproducer without alpaka?

**Comment by SimeonEhrig** [comment]
> If this happens in CUDA 11.6 we should open a NVIDIA bug report. Can you create a reproducer without alpaka?

That's correct but I'm not sure, if we can reproduce it. The math functions has a quiet complicated implementation :sweat_smile:

**Comment by sbastrakov** [comment]
I think for 1 trait it's reasonably possible to make a similar workflow to alpaka, just as `TAcc` use any type and "implement" the right trait for it

**Comment by sbastrakov** [comment]
I can make it after lunch if needed

**Comment by SimeonEhrig** [comment]
> I can make it after lunch if needed

Would be great.

**Comment by sbastrakov** [comment]
With a standalone reproducer I started with [this goldbolt piece](https://godbolt.org/z/dc1aKsfeT) but it surprisingly compiles. So i may have done smth wrong there, will check tomorrow. Also i didn't try to run it yet, could also trigger that `invalid device function` error.

**Comment by sbastrakov** [comment]
I made a standalone reproducer that consistently shows the behavior on the dev system with CUDA 11.0 (didn't try other versions). I am not exactly sure how to formulate all that, but I tried adding explanatory comments to the file, it's attached. I built it with the cmake as above, however that is probably not needed as I don't include alpaka, just reuse compilation flags.
[main.cpp.txt](https://github.com/alpaka-group/alpaka/files/8489805/main.cpp.txt)

**Comment by j-stephan** [comment]
Fixed by #1692.

**Comment by SimeonEhrig** [comment]
> I made a standalone reproducer that consistently shows the behavior on the dev system with CUDA 11.0 (didn't try other versions). I am not exactly sure how to formulate all that, but I tried adding explanatory comments to the file, it's attached. I built it with the cmake as above, however that is probably not needed as I don't include alpaka, just reuse compilation flags. [main.cpp.txt](https://github.com/alpaka-group/alpaka/files/8489805/main.cpp.txt)

The example looks fine. You can use the command `nvcc --std=c++17 --extended-lambda main.cu -o lambda_bug` with CUDA 11.6.2 to build the example without CMake.
